uProxy/uproxy-lib
# Big refactor to use require-style typescript + many cleanups

Addresses: https://github.com/uProxy/uproxy/issues/363
This is a significant code re-organization & rationalization of our build process.

Observations/Remarks: 
- This is rather a lot... sorry for the huge pull request! Hopefully no such big refactors of our build system  for a while now. 
- Using require in typescript was surprisingly easy (only had to spend 3 hours). 
- It makes dependencies more explicit and easier to follow
- The browserify stuff needed to actually use require statements in JS and manage building everything etc was surprisingly difficult to figure out (it interacted with some extra needed understanding of freedom, fixes to tests, and refinement of structure/location of files) (2 weeks)
- I think this is a LOT nicer, it's a now build structure with very few hacks (just one one's needed for using grunt).

Summary of work: 
- All typescript files now use require for importing typescript dependencies (one's which result in a require statement in the output code that is then linked by browserify)
- Separates interfaces for consumer/producer of handler-queues (implements: https://github.com/uProxy/uproxy/issues/657).
- Avoids test mock freedom in the for logging test (instead tests the interaction with freedom using freedom's APIs)
- Separates the `freedom-module-env` from the `freedom-core-env` typescript definitions. These are now two different definitions. This means that compilation of a typescript file for the `freedom-core-env` must now be done separately from a file that references  `freedom-module-env` (at least until freedom names the objects differently).
- Removed the need for writing/maintaining `*.d.ts` files. Now, `*.d.ts` files are now used only in the output `dev` and `dist` directories, and are generated programatically. 
- We now compile directly from `src` directory into `build/dev`. No more managing sym-links of typescript files in the build directory. 
- There is now a single build directory (like the main uProxy) which has `dev` and `dist` as subdirectories (no more root level dist). 
- Uses browserify as a JS linker, which is aware of the typescript compiled `require` statements. We no longer need to maintain the right web-page `<script src=...>` inclusions, nor chrome app script references, nor multiple freedom module script references: you now only need the single top level reference to the JS file. 
  - Dependencies used by sample apps are minimized: each freedom module is browserified into a single JS file: it contains only the files it uses, and is referenced from the freedom json manifest file. 
  - For spec-files, freedom-modules, and top-level sample-app files, each file needs to go through a browserification process. Browserify produces a `*.static.js` file that has all dependencies included. This automatically pulls in all the required files. 
  - Typescript compilation freedom modules must now be separately from compilation of a file in the core environment because they have different types for the (different, but named the same) `freedom` objects. 
- Move to using a new `install_dev_deps.sh` script: Given that tsd needs to be run to setup typescript definitions, and npm install scripts can't be selective for dev-dependencies (which means that you don't want to use an npm scripts for this), I've added in a little bash script for installing dev dependencies (it runs both the npm install, the appropriate tsd command, and tsc for build tools used by the gruntfile).

TESTED: 
- grunt test
- Setup on drone.io, shippable, and travis
- Ran sample apps & interactively tested them

TODO (in followup pull requests): 
- Make sure that uproxy-networking can use it: that means require-style version of the uproxy-networking repo.
- update coding guide & document some of the gotacha's
- Support sourcemaps
- Experiment with gulp instead of grunt.
- Think of a better name for some of the freedom typings & move freedom typings back to freedom-typescript-api repo



[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy-lib/100)
